### ISSUE TITLE
Support for enum type in CF 2.0

### ISSUE THREAD

**Issue by MTG-Formats** [cf-convention/CF-2#10]
CF 1.x uses an integer type with "flag_values" and "flag_meanings" attributes to create enumerated variables with self-describing flag values.

netCDF4 has added the enum type which also allows the user to create enumerated variables with self-describing flag types, but with a simpler syntax and greater re-usability. For instance, a boolean enum type may be defined once at the root level and then used consistently throughout the netCDF dataset.

Will CF 2.0 allow/support the use of the enum data type?

**Comment by MaartenSneepKNMI** [comment]
There is a third component in the flags system, and that is flag_masks. This is sometimes essential for the correct interpretation of the values in (processing) flags. For pure enumerations your proposal makes sense, but for bit-level flags (where an integer is used as a bit array), I think there are severe limitations of using an enumerated type.

**Comment by MTG-Formats** [comment]
I agree fully. We will make use of the flag_masks attribute to compress information such as when setting a series of boolean quality flags for each pixel in an image.

**Comment by JohnLCaron** [comment]
One sees a lot of code tables in BUFR and sometimes GRIB. The canonical
example might be land classification, where you have, say, 10 categories
and millions of data points. The enum data type allows you to store the
data as bytes, and store the 10 category descriptions once, with library
support for mapping the data value to the description. Essentialy this is
string valued data from a controlled vocabulary.

Flag values appear to be compact representations of phenomena that are
associated. This could be split into several fields and associated in a
structure ? Not sure of use case but I agree with Maarten its not a good
fit to the enum type.

The CDM data model supports netcdf-4 enum types, and I think they are a
good addition if others agree.

**Comment by MaartenSneepKNMI** [comment]
I agree that there are certainly good uses for an enumerated data type. The point I wanted to make is that there are other use cases that currently use the flag_values, flag_meanings and flag_masks that will continue to do so because the mapping to an enumerated type is awkward at best. So rather than one replacing the other, I think it is more a matter of adding support for enums.

**Comment by JonathanGregory** [comment]
Dear Tim et al.

If the enumerated data type had existed at the time the flags part of CF was introduced, quite probably we would have used it. However, if we introduce this new feature into the CF convention now, it will be an alternative to the existing one which will complicate software which has to read and write CF data, since both will need to be supported. Moreover some software which is not modified will be unable to read new data. We have agreed that backward incompatibility is sometimes the best choice, but I don't think we should make this change, even though it might be more elegant if we were starting from scratch, unless there is a strong reason why the existing convention is unsuitable or inadequate.

Best wishes

Jonathan

**Comment by graybeal** [comment]
I'm trying to puzzle something out. The original question was "Will CF 2.0 allow/support the use of the enum data type?"

So, the conformance checker won't complain if you don't use those attributes, right? -- I could put my flags metadata somewhere else, they just wouldn't be computer-interoperable. Similarly with NetCDF 4, where many users will undoubtedly come to use the enum convention for many of their files. I can't imagine we'd make the validation checkers find and flag enum use.  I also can't imagine that all those old pieces of unmodified software will happily read these nice new NetCDF4 files (that have enums, and many other features as well) -- that seems to me a red herring. (If I'm wrong about that, the rest of this mail is moot.)

Assuming that old netCDF software either will be upgraded or deprecated, there will be some features like enums in netCDF 4 that could map to existing/desired features/implementations in CF. Enums are discussed with some hope in the white paper, leaving some uncertainty:
   http://www.unidata.ucar.edu/software/netcdf/papers/nc4_conventions.html#Use_of_Enumerations_8762928607962679 
The paper speaks later to Jonathan's point:
> In general, it seems wise to avoid replacing an existing adequate convention with a better alternative convention that uses netCDF-4 features unless there is some significant advantage to the new convention, because applications will have to accept both the old and new conventions.

In this case a significant advantage is proposed, namely, enums can be globally or group-wide re-usable, in multiple contexts. They are also much more widely recognized.

If the options for new features like enums in CF2 are 'ignore', 'prohibit', or 'provide an alignment', I'm not sure which answer is advocated here. I think we have:
* ignore: users can use enums per NetCDF 4, but they won't be seen as contributing to CF2 interoperability; I would call that "allow" 
* prohibit: users can not use NetCDF 4 enums at all to describe flags; I think this is what Jonathan's answer says
* provide an alignment: This would qualify as 'support', and enums seem a case worth discussing further: maybe mapping from enums to flag values and flag_meanings (_not_ masks!), or declaring them a compatible option, is within the scope of manageable complexity? Especially if CF2 allows groups. 

If that summary is right, then my preferred answer to the question is "CF 2.0 will at least allow (through silence) the use of the enum data type, and could conceivably support it, if a low-cost, high-benefit alignment with the existing attributes is found."

John

**Comment by MTG-Formats** [comment]
If the CF 2.0 were to allow enums I think it would be better if it stated they are allowed but are unsupported rather than to allow them through silence. This would avoid the problem I have with trying to apply CF 1.6 to our netCDF-4 enhanced datasets. As an example, the conventions are silent on the subject of groups but also doesn't specify that the datasets must use a flat structure, and therefore leaves things ambiguous.

**Comment by mikeggrant-pml** [comment]
I agree enums would be a useful addition (the example of classification values is a good one).  While it sounds like flag_* have been used for this in the past, the enum type is a better fit.  It's also clear that the flags_* are best for bitmask-style flagging, and the name even implies this.

Could we add enums to the convention as the recommended way to represent enumerated classes and alter the flags_* section to recommend it for bitmasks without prohibiting its use for enumerations?  This wouldn't invalidate any previous data but would lead towards a clearer and useful distinction.

**Comment by JonathanGregory** [comment]
According to https://pypi.python.org/pypi/netCDF4, the netCDF4 python package currently doesn't support enum (or opaque) data types. Since python is quite commonly used to analyse CF data, that strikes me as a practical argument against allowing enum in CF for the moment.

I agree of course that enum is logically equivalent to using flag attributes to encode strings, and you might argue that it is nicer. However I see a couple of difficulties in mapping it to CF. (1) We use flag attributes sometimes to encode string-valued data variables such as `area_type`. For this to work properly, the translation of the data values into strings has to be provided; the enum CDL type definition does provide a translation between values and CDL strings, but between values and symbols used in CDL in the variable data. (2) The translation should be closely associated with the data variable, because variables may be copied between files. With the current convention, this naturally happens because the translation is stored in the flag attributes of the data variable. I am not sure how this would work with enums.

Jonathan

**Comment by MTG-Formats** [comment]
Re: Catering for netCDF-4 python 
Do the authors of the netCDF4 python package plan on having full support for netCDF4 in a future release, or will it forever be lacking enum suppport? Perhaps by including enum support in the CF conventions, we will see enum support arrive in the python package sooner?  ;-)  My feeling is that the CF conventions should be based on the netCDF-4 specification rather than the limited implementation of the specification in any given tool. There could be warnings provided that certain aspects may not be fully supported by certain current releases of popular  tools.

**Comment by MaartenSneepKNMI** [comment]
Be careful: some features are _only_ supported in the base C library, perhaps the Fortran interface, but not in the C++ interface. I'm not sure about enumerated data types, but a colleague is currently struggling with compound data types, and he is _not_ happy. Even in the C interface, support is flakey at best. 

Compound data types open a can of worms that in my opinion require a _very good_ use case, and careful consideration. Perhaps for CF2.1, but maybe not for 2.0.

**Comment by MTG-Formats** [comment]
Maybe this is an issue that should migrate to its own thread - whether CF 2.0 should be based on the specified netCDF-4 enhanced model or the currently executed netCDF-4 enhanced functionality available in the most commonly used tools and libraries? When specifying our netCDF-4 products we've made an implicit assumption that the tools would relatively quickly catch up with the model, but from reading comments here, this might be a bit optimistic.

**Comment by JonathanGregory** [comment]
For the CF convention we should be conservative about the use of netCDF features, because we want CF-netCDF data to be easily read and written by as large a community as possible. I think this applies to CF 2.0 just as much as before, since the purpose of CF is not changing. CF 2.0 offers an opportunity to clean up some things, if necessary (with caution) in a backward-incompatible way, but we still want a widely usable convention. Jonathan

**Comment by dblodgett-usgs** [comment]
@MTG-Formats and @JonathanGregory -- is this issue worth migrating to the cf-conventions repository? Hoping to limit the sprawl and archive things as needed.

**Comment by MTG-Formats** [comment]
I think it would be useful for the cf-conventions  to have a position on the use of new user types like enums. The conventions should either specify how to use them, discourage them, or state an indifference to their use. Reason for this is that it has never been clear whether a NetCDF product with groups can be considered CF-compliant as the CF conventions ,made no mention of groups. Having a clear position would avoid the same  problem for NetCDF products using enums or VLENs, etc.

**Comment by JonathanGregory** [comment]
I would suggest starting new issue rather than migrating it. Arguments in favour of allowing the new features should be made case-by-case, as was done for groups and is being done for strings.

**Comment by dblodgett-usgs** [comment]
ok. Do one of you want to open those issues in the repo? Should we maybe transfer this over there and close it just to have it over there for the record?

**Comment by MTG-Formats** [comment]
I don't see a transfer option so if somebody could transfer this issue across  - which was originally on the use of enumerated types - then I will open another asking that the conventions have a position on new features that aren't explicitly already covered by the document.
